plugins/bookmarks.js

```diff
@@ -367,13 +367,75 @@ window.plugin.bookmarks.loadStorageBox = function() {
     }
     // If portal isn't saved in bookmarks: Add this bookmark
     else{
-      // Get portal name and coordinates
-      var p = window.portals[guid];
-      var ll = p.getLatLng();
-      plugin.bookmarks.addPortalBookmark(guid, ll.lat+','+ll.lng, p.options.data.title);
+      window.plugin.bookmarks.addPortalBookmarkByGuid(guid, true);
     }
   }
 
+  /**
+   * Adds a portal to the default bookmark folder.
+   *
+   * @param {L.circleMarker} marker - As enhanced when added to
+   * window.portals.
+   * @param {boolean} doPostProcess - Whether additional post processing
+   * should be done after the bookmark was added.  E.g., saving to local
+   * storage, refreshing the widget, and running hooks.  If part of a batch
+   * update, this should probably be false.
+   */
+  window.plugin.bookmarks.addPortalBookmarkByMarker = function(marker, doPostProcess) {
+    const guid = marker.options.guid;
+    const label = marker.options.data.title;
+    const ll = marker.getLatLng();
+    const latlng = `${ll.lat},${ll.lng}`;
+    const ID = window.plugin.bookmarks.generateID();
+
+    window.plugin.bookmarks.bkmrksObj['portals'][window.plugin.bookmarks.KEY_OTHER_BKMRK]['bkmrk'][ID] = {
+      guid: guid,
+      latlng: latlng,
+      label: label,
+    };
+
+    if (doPostProcess) {
+      window.plugin.bookmarks.saveStorage();
+      window.plugin.bookmarks.refreshBkmrks();
+      window.runHooks('pluginBkmrksEdit', {
+        target: 'portal',
+        action: 'add',
+        id: ID,
+        guid: guid,
+      });
+      console.log(`BOOKMARKS: added portal ${ID}`);
+    }
+  }
+
+  /**
+   * Adds a portal to the default bookmark folder.
+   *
+   * @param {string} guid - The GUID of the portal.
+   * @param {boolean} doPostProcess - Whether additional post processing
+   * should be done after the bookmark was added.  E.g., saving to local
+   * storage, refreshing the widget, and running hooks.  If part of a batch
+   * update, this should probably be false.
+   */
+  window.plugin.bookmarks.addPortalBookmarkByGuid = function(guid, doPostProcess) {
+    const marker = window.portals[guid];
+    if (marker) {
```

Comment:
So it is possible that the portal will not be found and adding a bookmark will be quietly ignored?
In the old function the addition happens anyway. But whether it is good or bad is a question

Reply:
In the previous paths, the portal was looked prior to building the details.

In my edit, that would be lines 370 and 1180 in the original code.  The first is the same lookup, unchecked.  And the second is the `format` callback for the *Portals list* dialog, in which case, the marker is actually available in that context (you can tell it is a marker because it has the `getLatLng()` method).  It could actually make sense to replace that with a call to `addPortalBookmarkByMarker(portal, true);`

By definition, for something to be displayed on the map, there must be an existing `L.circleMarker()` in `window.portals`.  And both the sidebar and *Portals list* use that as a source to populate themselves.  So lack of checking in existing code is not an issue.

But, for any future uses, who knows where the guid may come from?  Say, someone wrote something to turn all saved key counts into bookmarks.  It is likely some of those would not be currently loaded because they're not in view.  So, yes, in that case, the request would be silently dropped.

The only other piece of code I found that does anything interesting when a portal cannot be found by guid is https://github.com/IITC-CE/ingress-intel-total-conversion/blob/2b47ad49356ab8da489d3bbff5d4b029f542f335/plugins/debug-raw-portal-data.js#L29 where all it does is a `console.warn`.

Best I can tell, nothing else uses this information to update user data.

Would something like:
```javascript
else {
  raise new Error(`Could not find portal information for guid "${guid}"`);
}
```
be appropriate/sufficient?

Reply:
Yes, let's do that:
1. Replace what you mentioned with addPortalBookmarkByMarker
2. Add raise new Error